metabase/metabase
# feat(sdk): support multiple instances of interactive questions with isolated redux stores

Closes https://github.com/metabase/metabase/issues/44670

### Description

Injects an isolated Redux store provider for each interactive questions, using the initial state from the root store. This makes it so that each interactive questions can operate independently, with independent query, filter and error states.

### Why?

The query builder is currently designed to operate on only one query at a time, e.g. one card, one query results, one error pages. With the SDK, we can have many queries at once. 

This is a hacky solution, but it is preferable to refactoring the entire query builder to support multiple instances of queries. I estimate multiple weeks of work if we go with refactoring the QB. I assume that most SDK users in the beginning won't have more than 20 interactive questions in a single page, so the additional Redux store's memory footprint is negligible. In addition, we still rely on a single root store for every static questions and static dashboards.

### How to verify

- Checkout the `test-isolated-redux-store` branch in the Shoppy app. Note that this is based on Vite (https://github.com/metabase/shoppy/pull/44) - please edit your `.env` and replace `REACT_APP` with `VITE_APP`, then run with `yarn dev`
- Navigate to the kitchen sink endpoint `/dev`
- In the `KitchenSink.tsx` page, there should be multiple interactive questions. You can add a couple more questions there.
- Each interactive questions must be able to operate independently.

### Caveat

- The initial state of the isolated store is inherited from the root redux store, therefore it inherits the authentication state (e.g. refresh token, login states) from when useMemo captures the initial state.

### Demo

https://github.com/metabase/metabase/assets/4714175/d8016dda-989a-43e1-91ed-a589d6e1a9c8

### Checklist

- [x] Tests have been added/updated to cover changes in this PR - to fix InteractiveQuestion tests
